### PR TITLE
Fix EZP-29375: REST API doesn't support multiple relation types

### PR DESCRIPTION
**Pull request by blankse**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29375](https://jira.ez.no/browse/EZP-29375)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.7`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

If a content has a relation with e. g. COMMON and EMBED relation type, the REST API can't load it.

Following exception is thrown:
```
Unknown relation type 3
```

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by blankse** [comment]
@andrerom Can you review it?

**Comment by andrerom** [comment]
For me it looks pretty good, indeed type can be set of several. But added some folks that have worked on REST recently to have a deeper review ;)

File: eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestRelation.php [hunk 1]
```diff
@@ -89,17 +89,25 @@ public function visit(Visitor $visitor, Generator $generator, $data)
...
-        switch ($relationType) {
-            case RelationValue::COMMON:
-                return 'COMMON';
-            case RelationValue::EMBED:
-                return 'EMBED';
-            case RelationValue::LINK:
-                return 'LINK';
-            case RelationValue::FIELD:
-                return 'ATTRIBUTE';
+        if ($relationType > 15 || $relationType < 1) {
```
Comment (alongosz) [inline eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestRelation.php]: That's not the best choice maintenance-wise :-)
Please move it before `return` statement and change to `if (empty($relationTypeList))`

Comment (blankse) [reply eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestRelation.php]: I changed it

**Comment by blankse** [comment]
@alongosz I created a test for relations.

There I noticed the following:
If I create a content with the REST API, the new stack creates multiple relations. In the legacy stack it is one relation. So I can't test this scenario...

Should I look into it and change this behavior?

**Comment by andrerom** [comment]
I'm wondering if that was on purpose or not. how does legacy behave when there is several field relations, then it can't really merge them can it?

File: eZ/Bundle/EzPublishRestBundle/Tests/Functional/RelationTest.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      <languageCode>eng-GB</languageCode>
+      <fieldValue>
+        <value key="destinationContentId">1</value>
+      </fieldValue>
+    </field>
+  </fields>
+</ContentCreate>
+XML;
+        $testContent = $this->createContent($xml);
+        self::assertEquals('EMBED,LINK,ATTRIBUTE', $testContent['CurrentVersion']['Version']['Relations']['Relation'][0]['RelationType']);
```
Comment (alongosz) [inline eZ/Bundle/EzPublishRestBundle/Tests/Functional/RelationTest.php]: Hmm, there's design problem here. Comma separated list of values does not conform to XML standard. It should be well formed XML, so if a `Relation` has `RelationType` element, then it should be split into three `Relation` elements. My tests show that this works (see my example above).

**Comment by blankse** [comment]
@andrerom I tested it in the legacy admin interface. Only common, embed and link are merged. Field relations have own relations.

@alongosz As I wrote in my last comment, the new stack make multiple relations. Legacy merge it in one relation.

Should I change the test to the current behavior?

**Comment by alongosz** [comment]
> the new stack make multiple relations. Legacy merge it in one relation.

Ah, ok. Now I get it. 

> Should I change the test to the current behavior?

Yes, unfortunately current behavior with XML normalized into multiple relations is the right one. In this case is the fix still needed for this repository? As said, for me it worked without an error, but indeed on a new stack.

**Comment by blankse** [comment]
@alongosz If you want support legacy, it is needed. Because merged relations run in a RuntimeException.

File: eZ/Bundle/EzPublishRestBundle/Tests/Functional/TestCase.php [hunk 3]
```diff
@@ -134,24 +134,24 @@ public function createHttpRequest($method, $uri, $contentType = '', $acceptType
         return $request;
     }
 
-    protected function assertHttpResponseCodeEquals(HttpResponse $response, $expected)
+    protected static function assertHttpResponseCodeEquals(HttpResponse $response, $expected)
```
[the remark this answers is not in the document]
Comment (blankse) [reply eZ/Bundle/EzPublishRestBundle/Tests/Functional/TestCase.php]: Yes, I noticed this when creating the test. This function is only used static. Should I put this changes in a single commit or revert?

[the remark this answers is not in the document]
Comment (blankse) [reply eZ/Bundle/EzPublishRestBundle/Tests/Functional/TestCase.php]: I reverted it

File: eZ/Publish/Core/REST/Client/Input/Parser/Relation.php [hunk 4]
```diff
@@ -72,18 +72,30 @@ public function parse(array $data, ParsingDispatcher $parsingDispatcher)
...
-                return Values\Content\Relation::LINK;
-            case 'FIELD':
-                return Values\Content\Relation::FIELD;
+        $stringTypeList = explode(',', strtoupper($stringType));
+        $relationType = 0;
+
+        foreach ($stringTypeList as $stringTypeValue) {
+            switch ($stringTypeValue) {
+                case 'COMMON':
+                    $relationType += Values\Content\Relation::COMMON;
```
[the remark this answers is not in the document]
Comment (blankse) [reply eZ/Publish/Core/REST/Client/Input/Parser/Relation.php]: I changed it to `|=`

**Comment by blankse** [comment]
I changed the test to the behavier of the new stack

**Comment by blankse** [comment]
@andrerom @alongosz I changed the test to the behavier of the new stack

**Comment by blankse** [comment]
Ping @andrerom @alongosz

**Comment by blankse** [comment]
@andrerom So it can be merged?

**Pull request merged**